osquery/logger/plugins/aws_log_forwarder.h

```diff
@@ -0,0 +1,301 @@
+    }
+
+    s = makeAWSClient<Client>(client_);
+    if (!s.ok()) {
+      return s;
+    }
+
+    return internalSetup();
+  }
+
+ private:
+  /// Dumps the specified batch to the given stream
+  std::ostream& dumpBatch(std::ostream& stream, const Batch& batch) {
+    std::size_t index = 0;
+
+    for (auto it = batch.begin(); it != batch.end(); it++) {
+      const auto& record = *it;
+      const auto& raw_buffer = record.GetData();
+
+      std::string buffer(
+          reinterpret_cast<const char*>(raw_buffer.GetUnderlyingData()),
+          raw_buffer.GetLength());
+      stream << "Record #" << index << ": " << buffer;
+
+      if (std::next(it, 1) != batch.end()) {
+        stream << "\n";
+      }
+    }
+
+    return stream;
+  }
+
+  /// Consumes the specified log records generating one or more batches
+  BatchList consumeDataAndGenerateBatches(
+      std::vector<std::string>& discarded_records,
+      const std::string& log_type,
+      std::vector<std::string>& log_data) {
+    BatchList batch_list;
+
+    Batch current_batch;
+    std::size_t current_batch_byte_size = 0U;
+
+    for (auto& record : log_data) {
+      // Initialize the line and make sure we are still within protocol limits
+      Status status = appendLogTypeToJson(log_type, record);
+      if (!status.ok()) {
+        // To achieve behavior parity with TLS logger plugin, skip non-JSON
+        // content
+        LOG(ERROR) << name_ << ": The following log record has been discarded "
+                               "because it was not in JSON format: "
+                   << record;
+
+        continue;
+      }
+
+      std::size_t record_size = record.size();
+      if (appendNewlineSeparators()) {
+        ++record_size;
+      }
+
+      if (record_size >= getMaxBytesPerRecord() ||
+          record_size >= getMaxBytesPerBatch()) {
+        discarded_records.push_back(std::move(record));
+        continue;
+      }
+
+      // Complete the current batch if it's full
+      if (current_batch_byte_size + record_size >= getMaxBytesPerBatch() ||
+          (current_batch.size() >= getMaxRecordsPerBatch())) {
+        batch_list.push_back(current_batch);
+
+        current_batch.clear();
+        current_batch_byte_size = 0U;
+      }
+
+      // Initialize and store the new log record
+      auto buffer = Aws::Utils::ByteBuffer(
+          reinterpret_cast<unsigned char*>(&record[0]), record_size);
+
+      if (appendNewlineSeparators()) {
+        buffer[record_size - 1] = '\n';
+      }
+
+      RecordType aws_record;
+      initializeRecord(aws_record, buffer);
+
+      current_batch.emplace_back(std::move(aws_record));
+      current_batch_byte_size += record_size;
+    }
+
+    if (!current_batch.empty()) {
+      batch_list.push_back(current_batch);
+    }
+
+    log_data.clear();
+    return batch_list;
+  }
+
+ protected:
+  /// Sends the specified data in one or more batches, depending on the log size
+  Status send(std::vector<std::string>& log_data,
```

Comment:
nit, this method is fairly complex, any way to break out the inner loop?

Reply:
Done, I've split it in two